Code/GraphMol/RascalMCES/RascalResult.h

```diff
@@ -0,0 +1,176 @@
+//
+// Copyright (C) David Cosgrove 2023
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+
+// A class to hold the results of a RASCAL MCES determination
+// between 2 molecules.  Contains the bonds and atoms that
+// correspond between the molecules, and also a SMARTS pattern
+// defining the MCES.
+//
+#include <RDGeneral/export.h>
+
+#ifndef RASCALRESULT_H
+#define RASCALRESULT_H
+
+#include <vector>
+
+#include <GraphMol/ROMol.h>
+
+namespace RDKit {
+
+namespace RascalMCES {
+
+class RDKIT_RASCALMCES_EXPORT RascalResult {
+ public:
+  RascalResult(const RDKit::ROMol &mol1, const RDKit::ROMol &mol2,
+               const std::vector<std::vector<int>> &adjMatrix1,
+               const std::vector<std::vector<int>> &adjMatrix2,
+               const std::vector<unsigned int> &clique,
+               const std::vector<std::pair<int, int>> &vtx_pairs, bool timedOut,
+               bool swapped, double tier1Sim, double tier2Sim,
+               bool ringMatchesRingOnly, bool singleLargestFrag,
+               int minFragSep);
+  // For when the tier[12]Sim didn't hit the threshold, but it
+  // might be of interest what the estimates of similarity were.
+  RascalResult(double tier1Sim, double tier2Sim);
+
+  RascalResult(const RascalResult &other);
+
+  RascalResult(RascalResult &&other) = default;
+
+  ~RascalResult() = default;
+
+  RascalResult &operator=(const RascalResult &other);
+
+  RascalResult &operator=(RascalResult &&other) = default;
+
+  // Cut the result down to the single largest fragment.  This is
+  // irrecoverably destructive.
+  void largestFragOnly();
+  void largestFragsOnly(unsigned int numFrags = 2);
+  void trimSmallFrags(unsigned int minFragSize = 3);
+
+  std::shared_ptr<RDKit::ROMol> getMol1() const { return d_mol1; };
+
+  std::shared_ptr<RDKit::ROMol> getMol2() const { return d_mol2; };
```

Comment:
I think we should try in the future (after merging this PR) to remove the need to copy the input molecules when we call the RascalMCES code. This would mean getting rid of the `d_mol1` and `d_mol2` shared pointers in `RascalResult.h`. 

We probably can't do that right away, but if we remove these two functions from the API (and the python wrappers) we can ensure that we'll be able to do it at some point in the future without breaking code. As far as I can tell, these two functions aren't used anywhere anyway.

Reply:
Some parts of the results, like the SMARTS of the match and the scores used in sorting results, are evaluated lazily.  The underlying molecules are therefore needed.  I guess that's a tradeoff that's open for discussion.  Some of the MCES manipulation functions, like imposing a minimum size of fragment that's retained, or cutting it down to the single largest fragment also require the underlying molecules to rebuild the SMARTS, for example.
I like the fact that the result can stand independently of the input set, but I'm not sure I can come up with a convincing use case off the top of my head.  I suppose it is easier to add these access functions later if they are needed than to remove them if they're not.

Reply:
> I suppose it is easier to add these access functions later if they are needed than to remove them if they're not.

yeah, that was my thinking